cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+    }
+
+    // case 5
+    Err(user_error(
+        "No remote specified, and no 'gerrit' remote was found",
+    ))
+}
+
+/// Determine what Gerrit ref and remote to use. The logic is:
+///
+/// 1. If the user specifies `--remote-branch branch`, use that
+/// 2. If the user has 'gerrit.default-remote-branch' configured, use that
+/// 3. Otherwise, bail out
+fn calculate_push_ref(
+    config: &UserSettings,
+    remote_branch: Option<String>,
+) -> Result<String, CommandError> {
+    // case 1
+    if let Some(remote_branch) = remote_branch {
+        return Ok(remote_branch);
+    }
+
+    // case 2
+    if let Ok(branch) = config.get_string("gerrit.default-remote-branch") {
+        return Ok(branch);
+    }
+
+    // case 3
+    Err(user_error(
+        "No target branch specified via --remote-branch, and no 'gerrit.default-remote-branch' \
+         was found",
+    ))
+}
+
+pub fn cmd_upload(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    upload: &UploadArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let revisions: Vec<_> = workspace_command
+        .parse_union_revsets(ui, &upload.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+    if revisions.is_empty() {
+        writeln!(ui.status(), "No revisions to upload.")?;
+        return Ok(());
+    }
+
+    if revisions
+        .iter()
+        .any(|commit| commit.id() == workspace_command.repo().store().root_commit_id())
+    {
+        return Err(user_error("Cannot upload the virtual 'root()' commit"));
+    }
+
+    workspace_command.check_rewritable(revisions.iter().ids())?;
+
+    // If you have the changes main -> A -> B, and then run `jj gerrit upload -r B`,
+    // then that uploads both A and B. Thus, we need to ensure that A also
+    // has a Change-ID.
+    // We make an assumption here that all immutable commits already have a
+    // Change-ID.
+    let to_upload: Vec<Commit> = workspace_command
+        .attach_revset_evaluator(
+            // I'm unsure, but this *might* have significant performance
+            // implications. If so, we can change it to a maximum depth.
+            Rc::new(RevsetExpression::Difference(
+                // Unfortunately, DagRange{root: immutable_heads, heads: commits}
+                // doesn't work if you're, for example, working on top of an
+                // immutable commit that isn't in immutable_heads().
+                Rc::new(RevsetExpression::Ancestors {
+                    heads: RevsetExpression::commits(
+                        revisions.iter().ids().cloned().collect::<Vec<_>>(),
+                    ),
+                    generation: jj_lib::revset::GENERATION_RANGE_FULL,
+                    parents_range: jj_lib::revset::PARENTS_RANGE_FULL,
+                }),
+                workspace_command.env().immutable_expression().clone(),
+            )),
+        )
+        .evaluate_to_commits()?
+        .try_collect()?;
```

Comment:
This expression should use a "range" (`x..y`). Something like `workspace_command.env().immutable_expression().range(&RevsetExpression::commits(...))`